### DISCUSSION TITLE
Allocating Memory over then  2G bytes

### DISCUSSION THREAD

**Discussion by aharondavid** [emscripten-core/emscripten#20287] [Q&A]
To overcome on allocating memory on the heap over then 2G, as I understand,  I have the following options:
1. Compiling the App with MAXIMUM_MEMORY=4g 
2. WebAssembly 64bit which is currently still in development phase 

Is an other options? 

I thought about a way to split the Shared Memory to a several process (each of them will contain max 2G bytes) , and a main process will access to those Shared Memory process.

**Comment by jspanchu** [comment]
@aharondavid I've had success allocating upto 6GB following these more or less:

1. If your project uses CMake, set `-DCMAKE_C_FLAGS:STRING="-sMEMORY64=1" -DCMAKE_CXX_FLAGS:STRING="-sMEMORY64=1"` from the command line during the first configure.
2. Enable `chrome://flags/#enable-experimental-webassembly-features` or use firefox nightly and run the application.
3. Link your application with `-sALLOW_MEMORY_GROWTH` and `-sMAXIMUM_MEMORY=nGB`. I've noticed that emscripten doesn't allow greater than 16GB in 64-bit mode.

> WebAssembly 64bit which is currently still in development phase

I'm not a browser/engine dev, but my favorite place to keep track of the progress is the https://github.com/WebAssembly/proposals page where you can see Memory64 is in Phase3 which means we could get it in stable browsers by the end of this year.

**Comment by aharondavid** [comment]
Setting MEMORY64=1 has a big effect on my App and maybe refactoring in some cases.
also MEMORY64 is not compatible with DISABLE_EXCEPTION_CATCHING=0
anyway, thank you! for your answer.

**Comment by sbc100** [comment]
> To overcome on allocating memory on the heap over then 2G, as I
> understand, I have the following options:
>
>    1. Compiling the App with MAXIMUM_MEMORY=4g
>    2. WebAssembly 64bit which is currently still in development phase
>
> Is an other options?
>

As far as I know those are your two options.   I don't know of any kind of
memory sharding options, at least no in emscripten.
